### DISCUSSION TITLE
Bun v0.0.56

### DISCUSSION THREAD

**Discussion by Jarred-Sumner** [oven-sh/bun#85] [Announcements]
## bun install

Today is an exciting day.  Introducing `bun install` – an incredibly fast npm client.

Before I get into perf numbers, consider this especially early. 
- workspaces not implemented yet
- `link:` not implemented yet
- `github:` not implemented yet
- `git:` not implemented yet
- Config file where you can set registry and auth tokens is not implemented yet. That means private packages won't really work yet (though it does read `$npm_config_registry`)
- The lockfile format may change a little in the coming weeks (which would invalidate existing lockfiles)
-  postinstall will not be supported (probably won't be [opt-in](https://github.com/npm/rfcs/pull/488) either).  `esbuild` and `turbo` get special treatment I'm calling "native bin linking". I plan to write a more generalizable proposal for this that other npm clients can adopt to make shipping binary executables on npm faster & simpler
- `node-gyp` isn't supported yet either, but that will be fixed.

### Linux 

 **100x faster npm install** for incremental installs. In this case, only react is missing (the cleanup resets node_modules), each package manager has an up-to-date lockfile, and packages have previously downloaded before (disk cache)

<img src="https://user-images.githubusercontent.com/709451/146924903-175dcfbd-c3a5-4a89-860b-c73178b35268.png"  />

20x faster than npm install when node_modules is empty. In this case, each package manager has an up-to-date lockfile and packages have also been downloaded before.

![image](https://user-images.githubusercontent.com/709451/147004342-571b6123-17a9-49a2-8bfd-dcfc5204047e.png)

When there are no changes, `bun install` is 100x faster than npm install

![image](https://user-images.githubusercontent.com/709451/147003624-82154661-42f9-4cd4-b09a-6b66c2c56ff6.png)

### macOS 

Note: currently macOS installs are single-threaded & parallelism may help when there are > 100 packages, so these numbers may improve later

80x faster npm install for incremental installs.

![image](https://user-images.githubusercontent.com/709451/147006020-1c4c79ce-269c-4418-96c8-999b0da103a1.png)

6x faster npm install when there is no node_modules folder

![image](https://user-images.githubusercontent.com/709451/146926382-d4038017-3bb5-4ba3-ab06-5f322025636a.png)

When there are no changes, 80x faster than npm install 

![image](https://user-images.githubusercontent.com/709451/147006817-9251ead6-68fc-47be-a777-fc1ad807fea0.png)

### Autocomplete

Thanks to @evanwashere, `bun add` has autocomplete for the top 10,000 npm packages. It also searches your shell's history to complete from.

![Screen_Recording_2021-12-19_at_11 32 08_PM](https://user-images.githubusercontent.com/709451/147006855-7e323039-b447-4849-aa44-07c93457372f.gif)

Fish completions:

![image](https://user-images.githubusercontent.com/709451/147006938-46b0e831-e32d-456c-9179-f362f847b851.png)

#### Why is it faster?

I will do a longer write up on why it's faster. Zig is part of that, but a lot of work went into data layout, the lockfile format, the https client, the manifest cache format, and a lot of other stuff. 

<hr /><em>This discussion was created from the release <a href='https://github.com/Jarred-Sumner/bun/releases/tag/bun-v0.0.56'>Bun v0.0.56</a>.</em>

**Comment by alirezabonab** [comment]
@Jarred-Sumner Is there an estimate for when bun will support workspaces?
